Exponential Sine Sweep

estearg opened the pull request:
To be able to characterize nonlinearity, getting each harmonic impulse response, an exponential sine sweep must be performed and an inverse filter applied to the signal in order to obtain the impulse response.
mataa_ESS_sync does that,  with zero phase at beginning and end, thus minimizing the magnitude spectrum ripple in the high-frequency extreme, and with phase synchronization between different order impulse responses. To reduce ripple even more, 1-octave fade-in and 1/24-octave fade-out Hanning windows are applied

Commit:
Added tool to perform Exponential Sine Sweep with phase synchronization at beginning and end and between different order impulse responses

mbrennwa commented on mataa_tools/mataa_ESS_sync.m:
```diff
@@ -0,0 +1,117 @@
+% > P = 8;																										% number of octaves down from Nyquist, plus 1 for fade-in
+% > T = 10;																										% approximate duration of swepp
+% > fs = 96000;																									% sampling frequency
+% > mataa_ESS_sync(P, T, fs, 0.22);																				% perform an ESS with 0.22 s latency
```
this is missing [h,t] = ...

mbrennwa commented on mataa_tools/mataa_ESS_sync.m:
```diff
@@ -0,0 +1,117 @@
+% > set(findall(gcf, 'type', 'axes'), 'xgrid', 'on', 'xminorgrid', 'on', 'ygrid', 'on', 'yminorgrid', 'on');	% make it easier to see the coordinates of a part of the response
+% > t0 = mataa_guess_IR_start(h, fs);																			% find the beginning of the linear response
+% > xlim([t0 t0+0.01]);																							% focus on the first 10 ms of the linear response
+% > tfirstecho = 5.0511-t0;																						% time of first reflexion, as seen in the plot, relative to beginning of response
```
Where does the 5.0511 value come from?

mbrennwa commented on mataa_tools/mataa_ESS_sync.m:
```diff
@@ -0,0 +1,117 @@
+function [h, t] = mataa_ESS_sync(P, T, fs, latency, cal)
```
I'd suggest to change the function name following the logic of the other MATAA functions. This is a measurement, so it should be mataa_measure_XXX. The measurement result is an impulse response (IR) and the harmonic distortion (HD). How about mataa_measure_IR_HD?

mbrennwa commented on mataa_tools/mataa_ESS_sync.m:
```diff
@@ -0,0 +1,117 @@
+%
+% > P = 8;																										% number of octaves down from Nyquist, plus 1 for fade-in
+% > T = 10;																										% approximate duration of swepp
+% > fs = 96000;																									% sampling frequency
```
I'd suggest to change this to 44100, which is available on most systems.

mbrennwa commented on mataa_tools/mataa_ESS_sync.m:
```diff
@@ -0,0 +1,117 @@
+h = shift(h, -floor(length(responseSignal) / 2) + 1);
+h = resize(h, length(responseSignal), 1);
+
+endfunction
```
This looks like GNU Octave only. Have you tested this with Matlab?

mbrennwa commented on mataa_tools/mataa_ESS_sync.m:
```diff
@@ -0,0 +1,117 @@
+n = [0:N]';
+M = round(N / (log(2 ^ P) * (2 ^ (P + 1))));
+
+disp("Note that if later mataa_IR_to_FR is called, it adds 112 dB to the magnitude");
```
I don't see the point of this message. I'd suggest to remove this to avoid confusion.

mbrennwa commented on mataa_tools/mataa_ESS_sync.m:
```diff
@@ -0,0 +1,117 @@
+endfor
+
+# ESS generation
+x = sin(2 .* pi .* M .* exp(n .* log(2 ^ P) ./ N));
```
why don't you use mataa_signal_generator?

mbrennwa commented on mataa_tools/mataa_ESS_sync.m:
```diff
@@ -0,0 +1,117 @@
+% DESCRIPTION:
+% Generates an exponential sine sweep containing an integer number of octaves down from the Nyquist frequency and lasting approximately T seconds, then measures the response signal and applies the inverse filter to it, thus obtaining the impulse response h.
+% All impulse responses of a non-linear system are phase-synchronized, and the magnitude spectrum ripple in the high-frequency extreme is minimized due to the sweep beginning and ending in phase zero.
+% A Hanning fade-in is applied for the first octave, so the flat spectrum is (P - 1) octaves long. Similarly, the last 1/24 octave is faded out with a Hanning window to reduce the ripple even more.
```
A Hanning HALF window at the end?

mbrennwa commented:
Thanks for making this available! However, I could not fully test the function, because I didn't completely understand the example. If you could change some minor things to the code (I added some notes to your code) I'd be happy to include it in main MATAA.

estearg commented:
Hello Matthias,

First of all, thank you very much for MATAA.

I'll modify the file according to all your comments on it.
But regarding MATLAB, I have no access to a computer running that
program. Perhaps once I finish making the changes --which will include
revising the Octave-only code-- you can test it on MATLAB.

Best regards,

Esteban

2017-12-24 1:15 GMT-03:00, mbrennwa <notifications@github.com>:
> Thanks for making this available! However, I could not fully test the
> function, because I didn't completely understand the example. If you could
> change some minor things to the code (I added some notes to your code) I'd
> be happy to include it in main MATAA.
>
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/mbrennwa/mataa/pull/3#issuecomment-353764269

mbrennwa commented:
Hi Esteban

Great! Don't worry too much about the Matlab testing. I also haven't touched Matlab for a few years now (no need for it). You could use Octave's "braindead" mode and enable "language-extension" warnings for some basic Matlab compatibility tests; that will capture most issues easily.

Thanks for contributing your great work to MATAA! Please also add a line to the help text saying that this code was contributed by you.

Thanks
Matthias

estearg commented:
Hello Matthias,

It was good to receive your comments, they made me revise the code and
read some more about exponential sine sweeps.

According to MathWorks'  website, it is now compatible with MATLAB.

The only suggestion I failed to follow is that on using
mataa_signal_generator; I wasn't able to find the kind of signal
needed.

Best regards,

Esteban


2017-12-29 20:34 GMT-03:00, mbrennwa <notifications@github.com>:
> Merged #3.
>
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/mbrennwa/mataa/pull/3#event-1404937884

mbrennwa commented:
Thanks! I also worked a bit on it in the meantime. I'll have to merge everything and hope I don't screw up too much. Where can I see your revised file?

estearg commented:
Hello Matthias,

Yesterday I pushed mataa_measure_IR_HD.m to branch ESS of my fork of
MATAA, but forgot to remove mataa_ess_sync.m.
Now it's done. It's at:
https://github.com/estearg/mataa/blob/ESS/mataa_tools/mataa_measure_IR_HD.m

Best regards,

Esteban

2017-12-30 13:21 GMT-03:00, mbrennwa <notifications@github.com>:
> Thanks! I also worked a bit on it in the meantime. I'll have to merge
> everything and hope I don't screw up too much. Where can I see your revised
> file?
>
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/mbrennwa/mataa/pull/3#issuecomment-354554492

mbrennwa commented:
Thanks! I combined your changes with my file, where I have also added a few more features to the function. I also changed a few calls to the signal package to the MATAA versions (to make the code independent of the signal package). Please take take a look at the file (https://github.com/mbrennwa/mataa/blob/master/mataa_tools/mataa_measure_IR_HD.m) and let me know if everything is ok.

I attached a screenshot of a measurement I did with the mataa_measure_IR_HD function and my laptop speakers + microphone, where I allowed the test signal to clip the input. This gave quite a bit of harmonics. The measurement and figures were done as in the EXAMPLE included in the mataa_measure_IR_HD function.

![farina_demo](https://user-images.githubusercontent.com/9806692/34461131-462cafe2-ee23-11e7-93e1-29a54ee9a58a.png)

estearg commented:
Hello Matthias,

It works as expected and the improvements are worthy. Besides, the
description is clearer now.

Sometimes I wonder what can be done to give more visibility to MATAA,
because it' a really very useful package. Thank you again for it.

Best regards,

Esteban

2017-12-31 8:09 GMT-03:00, mbrennwa <notifications@github.com>:
> Thanks! I combined your changes with my file, where I have also added a few
> more features to the function. I also changed a few calls to the signal
> package to the MATAA versions (to make the code independent of the signal
> package). Please take take a look at the file
> (https://github.com/mbrennwa/mataa/blob/master/mataa_tools/mataa_measure_IR_HD.m)
> and let me know if everything is ok.
>
> I attached a screenshot of a measurement I did with the mataa_measure_IR_HD
> function and my laptop speakers + microphone, where I allowed the test
> signal to clip the input. This gave quite a bit of harmonics. The
> measurement and figures were done as in the EXAMPLE included in the
> mataa_measure_IR_HD function.
>
> ![farina_demo](https://user-images.githubusercontent.com/9806692/34461131-462cafe2-ee23-11e7-93e1-29a54ee9a58a.png)
>
>
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/mbrennwa/mataa/pull/3#issuecomment-354598050

estearg commented:
Oh, and I like that clipping test you did.

2017-12-31 20:16 GMT-03:00, Esteban Lucchesi <lucchesi.esteban@gmail.com>:
> Hello Matthias,
>
> It works as expected and the improvements are worthy. Besides, the
> description is clearer now.
>
> Sometimes I wonder what can be done to give more visibility to MATAA,
> because it' a really very useful package. Thank you again for it.
>
> Best regards,
>
> Esteban
>
> 2017-12-31 8:09 GMT-03:00, mbrennwa <notifications@github.com>:
>> Thanks! I combined your changes with my file, where I have also added a
>> few
>> more features to the function. I also changed a few calls to the signal
>> package to the MATAA versions (to make the code independent of the signal
>> package). Please take take a look at the file
>> (https://github.com/mbrennwa/mataa/blob/master/mataa_tools/mataa_measure_IR_HD.m)
>> and let me know if everything is ok.
>>
>> I attached a screenshot of a measurement I did with the
>> mataa_measure_IR_HD
>> function and my laptop speakers + microphone, where I allowed the test
>> signal to clip the input. This gave quite a bit of harmonics. The
>> measurement and figures were done as in the EXAMPLE included in the
>> mataa_measure_IR_HD function.
>>
>> ![farina_demo](https://user-images.githubusercontent.com/9806692/34461131-462cafe2-ee23-11e7-93e1-29a54ee9a58a.png)
>>
>>
>> --
>> You are receiving this because you authored the thread.
>> Reply to this email directly or view it on GitHub:
>> https://github.com/mbrennwa/mataa/pull/3#issuecomment-354598050
>
